### PR TITLE
OGM-786 GridFS support

### PR DESCRIPTION
**Pull request by schernolyas**
https://hibernate.atlassian.net/browse/OGM-786

### CODE REVIEW

File: mongodb/src/test/java/org/hibernate/ogm/datastore/mongodb/test/gridfs/GridFSTest.java [hunk 1]
```diff
@@ -153,6 +164,8 @@ public void canRemoveEntityWithBlob() {
 			assertBlobAreEqual( photo.getContentAsBlob(), BLOB_CONTENT_1 );
 			assertThat( photo.getContentAsByteArray() ).isNotNull();
 			assertThat( photo.getContentAsByteArray() ).isEqualTo( BLOB_CONTENT_1 );
+			assertThat( photo.getContentAsString() ).isNotNull();
+			assertThat( photo.getContentAsString() ).isEqualTo( BLOB_STRING_CONTENT_1 );
```
Comment (DavideD) [inline mongodb/src/test/java/org/hibernate/ogm/datastore/mongodb/test/gridfs/GridFSTest.java]: I don't think these tests are right for this issue. They would pass even if you don't store the field using GridFS. Is there a way to check if the field has been store effectively using GridFS?

Also there is no need to asser

File: mongodb/src/test/java/org/hibernate/ogm/datastore/mongodb/test/gridfs/GridFSTest.java [hunk 2]
```diff
@@ -153,6 +164,8 @@ public void canRemoveEntityWithBlob() {
 			assertBlobAreEqual( photo.getContentAsBlob(), BLOB_CONTENT_1 );
 			assertThat( photo.getContentAsByteArray() ).isNotNull();
 			assertThat( photo.getContentAsByteArray() ).isEqualTo( BLOB_CONTENT_1 );
+			assertThat( photo.getContentAsString() ).isNotNull();
```
Comment (DavideD) [inline mongodb/src/test/java/org/hibernate/ogm/datastore/mongodb/test/gridfs/GridFSTest.java]: This line is not needed, the  isEqualTo already take care of null values

Comment (schernolyas) [reply mongodb/src/test/java/org/hibernate/ogm/datastore/mongodb/test/gridfs/GridFSTest.java]: Ok. I will remove excess line.

File: mongodb/src/test/java/org/hibernate/ogm/datastore/mongodb/test/gridfs/GridFSTest.java (hunk 1 above)
Comment (schernolyas) [reply mongodb/src/test/java/org/hibernate/ogm/datastore/mongodb/test/gridfs/GridFSTest.java]: you right.  About your question. Yes. We can check GridFS. We can write additional information to meta.

File: mongodb/src/test/java/org/hibernate/ogm/datastore/mongodb/test/gridfs/GridFSTest.java [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+			photo.setContentAsByteArray( BLOB_CONTENT_1 );
+			photo.setContentAsString( BLOB_STRING_CONTENT_1 );
+			em.persist( photo );
+		} );
+
+		inTransaction( em -> {
+			MongoDatabase mongoDatabase = getCurrentDB( em );
+			GridFSBucket gridFSFilesBucket = GridFSBuckets.create( mongoDatabase, BUCKET_NAME );
+			MongoCursor<GridFSFile> cursor = gridFSFilesBucket.find().iterator();
+			for ( int i = 0; i < 3; i++ ) {
```
Comment (DavideD) [inline mongodb/src/test/java/org/hibernate/ogm/datastore/mongodb/test/gridfs/GridFSTest.java]: I don't understand why you are using a `for` loop in this scenario when you have a cursor.
What happens if the cursor returns 10 values but you only check the first 3?
Also, where the number 3 is coming from?

Comment (schernolyas) [reply mongodb/src/test/java/org/hibernate/ogm/datastore/mongodb/test/gridfs/GridFSTest.java]: I understood what you mean. Thank for 
 criticism and idea!

File: mongodb/src/test/java/org/hibernate/ogm/datastore/mongodb/test/gridfs/GridFSTest.java (hunk 1 above)
Comment (DavideD) [reply mongodb/src/test/java/org/hibernate/ogm/datastore/mongodb/test/gridfs/GridFSTest.java]: Yes, I guess... you need to make sure that it contains the values you expect. Does that asnwer your question?

Comment (schernolyas) [reply mongodb/src/test/java/org/hibernate/ogm/datastore/mongodb/test/gridfs/GridFSTest.java]: Hi @DavideD !

Yes,of course.

**Comment by DavideD** [comment]
I'm going to close this one.
After trying to refactor this PR I realized that this approach of using the annotation doesn't work well with our existing code and make it harder to have custom types.

I also decided that it's not helpful to assume that there are going to be other binary storages for now. This will simplify the code a lot,

Even if we go with this approach there are some cases there are not covered by tests and make the PR fails.

I will send a new PR when I have time. Currently, there is PREVIEW on my repository: https://github.com/DavideD/hibernate-ogm/tree/OGM-786-Preview

Thanks

**Pull request closed without merging**